File: backend/app/routers/lemon_squeezy.py
"""Lemon Squeezy 路由 —— /api/v1/credits/lemonsqueezy（阶段一：只建单，不发放）。

- GET  /status    公开：LS 链路是否可用 + 已配 Variant 的条目 id（不含任何密钥/凭据）
- POST /checkout  登录：服务端向 LS 建单，只返回 {checkout_url}

与 Paddle 的关系：本文件**不改动** Paddle 的任何行为。两条链路各自独立配置，
前端按 /status 决定某一条目走哪条；LS 未配置时结果与今天完全一致。

安全边界：
1. 客户端只能提交 pack_id 或 plan_id（二选一），模型 extra="forbid" —— 多传任何字段
   （price / credits / variant_id / custom 等）直接 422，不会被后端"顺带采纳"。
2. Variant ID、积分数、金额一律由服务端从配置解析，绝不采信客户端。
3. 本阶段不存在任何发放 Credits 的路径（webhook 尚未实现）。
"""

from __future__ import annotations

import logging
import os
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, ConfigDict

from app.services.auth_identity import get_verified_user_id
from app.services.credits_config import CREDIT_PACKS, MEMBERSHIP_PLANS
from app.services import lemon_squeezy_service as ls

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/credits/lemonsqueezy", tags=["credits"])


class LemonSqueezyCheckoutRequest(BaseModel):
    """只允许指定"买哪个条目"。任何其它字段（含价格/积分/Variant ID）一律拒收。"""
    model_config = ConfigDict(extra="forbid")

    pack_id: Optional[str] = None
    plan_id: Optional[str] = None


@router.get("/status")
async def lemonsqueezy_status():
    """供前端判断走哪条链路。只暴露布尔与条目 id，不回显任何凭据或 Variant ID。"""
    return {
        "enabled": ls.checkout_enabled(),
        "items": ls.configured_items(),
    }


@router.post("/checkout")
async def create_lemonsqueezy_checkout(req: LemonSqueezyCheckoutRequest,
                                       user_id: str = Depends(get_verified_user_id)):
    """为积分包或会员订阅创建 Lemon Squeezy Checkout（本端点不发放任何 Credits）。"""
    if bool(req.pack_id) == bool(req.plan_id):
        raise HTTPException(422, "必须且只能指定 pack_id 或 plan_id 之一")

    if req.pack_id:
        item = next((p for p in CREDIT_PACKS if p["id"] == req.pack_id), None)
        if item is None:
            raise HTTPException(404, "积分包不存在或未开放购买")
        item_id = item["id"]
        custom = {"user_id": user_id, "kind": "credit_pack",
                  "pack_id": item_id, "credits": str(item["credits"])}
    else:
        plan = next((p for p in MEMBERSHIP_PLANS if p["id"] == req.plan_id), None)
        if plan is None:
            raise HTTPException(404, "会员计划不存在或未开放订阅")
        item_id = plan["id"]
        custom = {"user_id": user_id, "kind": "membership",
                  "plan_id": item_id, "credits_per_month": str(plan["credits_per_month"])}

    if not ls.checkout_enabled():
        raise HTTPException(503, "lemonsqueezy_not_configured")
    variant_id = ls.variant_id_for(item_id)
    if not variant_id:
        # 商店/密钥配了但这一档没配 Variant：宁可 503 也不让前端拿到一个错商品的链接
        raise HTTPException(503, "lemonsqueezy_variant_not_configured")

    redirect_url = (os.environ.get("LEMONSQUEEZY_REDIRECT_URL") or "").strip() or None
    try:
        order = await ls.create_checkout(variant_id=variant_id, custom=custom,
                                         redirect_url=redirect_url)
    except ls.LemonSqueezyError as exc:
        logger.warning("Lemon Squeezy checkout failed for %s: code=%s status=%s",
                       user_id, exc.code or "-", exc.status or "-")
        raise HTTPException(502, exc.code or "lemonsqueezy_checkout_failed")

    return {"checkout_url": order["checkout_url"]}

File: frontend/src/lib/lemonSqueezy.ts
/**
 * Lemon Squeezy 前端适配层（阶段一）。
 *
 * 只做两件事：问后端"这条链路可用吗"，以及拿 checkout_url 后跳转到 Hosted Checkout。
 * 刻意不引入 Lemon.js / overlay SDK —— LS 官方支持纯跳转式 Hosted Checkout，
 * 少一个第三方全局脚本就少一类供应链与 CSP 问题。
 *
 * 前端永远不接触 API Key、Store ID、Variant ID：那些全在后端环境变量里。
 * 这里也**不**做任何"付款成功"判定：发放只认后端 webhook（阶段一尚未实现）。
 */

import { api } from '../config/api';
import { authFetchOptional } from '../api/http';

export interface LemonSqueezyStatus {
  enabled: boolean;
  /** 已配好 Variant ID 的条目 id（如 credits_200 / starter）；不在表里的仍走 Paddle。 */
  items: string[];
}

export async function fetchLemonSqueezyStatus(): Promise<LemonSqueezyStatus> {
  try {
    const res = await authFetchOptional<LemonSqueezyStatus>(
      `${api.base}/api/v1/credits/lemonsqueezy/status`,
    );
    return { enabled: res?.enabled === true, items: Array.isArray(res?.items) ? res.items : [] };
  } catch {
    // 探测失败一律当作"不可用"：购买链路退回 Paddle，绝不因为多一个 provider
    // 而把原本能点的按钮变成死的。
    return { enabled: false, items: [] };
  }
}

/**
 * 让后端建一张 Lemon Squeezy Checkout 并跳转过去。
 * 请求体只允许 pack_id 或 plan_id —— 价格/积分数/Variant ID 由后端解析，
 * 多传字段会被后端 422 拒收（extra="forbid"）。
 */
export async function openLemonSqueezyCheckout(
  body: { pack_id?: string; plan_id?: string },
): Promise<boolean> {
  const res = await authFetchOptional<{ checkout_url?: string }>(
    `${api.base}/api/v1/credits/lemonsqueezy/checkout`,
    { method: 'POST', body },
  );
  const url = res?.checkout_url;
  if (!url) return false;
  window.location.assign(url);
  return true;
}
